devito/ir/support/basic.py

```diff
@@ -529,9 +530,16 @@ def __hash__(self):
             (self.source, self.sink, self.source.timestamp == self.sink.timestamp)
         )
 
-    @property
+    @cached_property
     def function(self):
-        return self.source.function
+        if q_comp_acc(self.source.access) and not q_comp_acc(self.sink.access):
+            # E.g., `source=ab[x].x` and `sink=ab[x]` -> `a(x)`
+            return self.source.access.function_access
+        elif q_comp_acc(self.sink.access) and not q_comp_acc(self.source.access):
```

Comment:
nitpicking: keep same condition order so `not q_comp_acc(self.source.access) and q_comp_acc(self.sink.access)`